controller/src/main/java/io/pravega/controller/store/stream/records/SubscriberConfiguration.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream.records;
+
+import com.google.common.collect.ImmutableMap;
+import io.pravega.common.ObjectBuilder;
+import io.pravega.common.io.serialization.RevisionDataInput;
+import io.pravega.common.io.serialization.RevisionDataOutput;
+import io.pravega.common.io.serialization.VersionedSerializer;
+import lombok.Builder;
+import lombok.Data;
+import lombok.SneakyThrows;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * The configuration of a Stream Subscriber.
+ */
+@Data
+@Builder
+public class SubscriberConfiguration implements Serializable {
+
+    public static final SubscriberConfigurationSerializer SERIALIZER = new SubscriberConfigurationSerializer();
+    public static final SubscriberConfiguration EMPTY = new SubscriberConfiguration(0L, ImmutableMap.of());
+    private static final long serialVersionUID = 1L;
+    /**
+     * Time when this configuration was updated.
+     */
+    private final long updateTime;
+
+    /**
+     * Truncation Stream cut published by this subscriber
+     */
+    private final ImmutableMap<Long, Long> streamCut;
```

Comment:
please dont put the position/streamcut inside the subscriber record. since this record is actually written under the `StreamSubscribersRecord` this will be problematic on two dimensions:
1. with large number of subscribers, the size of this record can exceed the 1mb limit of kvt record if there are large number of segments in the streamcut. its better to divide and rule. 
2. this will mean StreamSubscribersRecord is upadted with every "report" from a subscriber which means the write conflicts can occur lot more whereby making optimistic concurrency undesirable.

my suggestion is to do as follows:
- StreamSubscribersRecord should only contain a list of subscriber ids. 
- create separate record for each subscriber for their reported position.

addSubscriber/removeSubscriber updates the StreamSubscribersRecord with the subscriber's id. 
then for each subscriber there is a separate key called say `subscriber_<subscriberId>`.
when subscriber with `id` reports its streamcut, its streamcut is updated in `subscriber_id` record. 

getSubscribers will first fetch the StreamSubscribersRecord  and then fetch all their individual records corresponding to their last reported positions. 

removesubscriber can perform batch update of StreamSubscribersRecord and subscriber_id record by updating the former and removing the later via an atomic kvt update.

Reply:
> 1. with large number of subscribers, the size of this record can exceed the 1mb limit of kvt record if there are large number of segments in the streamcut. its better to divide and rule.

The 1 MB limit would be exceeded even when we maintain separate records for subscribers and truncation stream-cuts. Assuming each 'Subscriber --> SC' map entry takes up 100 bytes, it would take 10K subscribers to hit that limit....which is a pretty large number.... Having Subscriber name as part of the same record only marginally impacts the table segment record limitation.....

> 2. this will mean StreamSubscribersRecord is updated with every "report" from a subscriber which means the write conflicts can occur lot more whereby making optimistic concurrency undesirable.

The downside of maintaining 2 separate records is the need to maintain consistency between them.
Optimistic concurrency concerns due to frequent update of truncation stream-cuts are very valid, but it has nothing to do with Subscribers being part of the same metadata, as subscriber addition is a much infrequent operation.

One way to deal with this can be to have a separate record in Stream Metadata per subscriber and update just that record on stream-cut updates.

Reply:
fixed